pkg/logentry/stages/replace.go

```diff
@@ -0,0 +1,240 @@
+package stages
+
+import (
+	"bytes"
+	"reflect"
+	"regexp"
+	"strconv"
+	"text/template"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+)
+
+// Config Errors
+const (
+	ErrEmptyReplaceStageConfig = "empty replace stage configuration"
+	ErrEmptyReplaceStageSource = "empty source in replace stage"
+)
+
+// RegexConfig contains a regexStage configuration
+type ReplaceConfig struct {
+	Expression string  `mapstructure:"expression"`
+	Source     *string `mapstructure:"source"`
+	Replace    string  `mapstructure:"replace"`
+}
+
+// validateReplaceConfig validates the config and return a regex
+func validateReplaceConfig(c *ReplaceConfig) (*regexp.Regexp, error) {
+	if c == nil {
+		return nil, errors.New(ErrEmptyReplaceStageConfig)
+	}
+
+	if c.Expression == "" {
+		return nil, errors.New(ErrExpressionRequired)
+	}
+
+	if c.Source != nil && *c.Source == "" {
+		return nil, errors.New(ErrEmptyReplaceStageSource)
+	}
+
+	if c.Replace == "" {
+		return nil, errors.New(ErrEmptyReplaceStageConfig)
+	}
+
+	expr, err := regexp.Compile(c.Expression)
+	if err != nil {
+		return nil, errors.Wrap(err, ErrCouldNotCompileRegex)
+	}
+	return expr, nil
+}
+
+func validateReplaceString(replace string) (string, int, error) {
+	replaceCount := -1
+	var replaceString string
+
+	r, err := regexp.Compile("(^\\$\\d+)(.*)")
+	if err != nil {
+		return "", 0, nil
+	}
+
+	match := r.FindStringSubmatch(replace)
+
+	if len(match) > 1 {
```

Comment:
I don't think this syntax is interesting what use case are we trying to cover ?

Reply:
This is to cover the following scenario. 
If there are multiple matches for a regex, how many matches a user would want to replace. Explained in the docs above, this part `replace: '$3test'`